extensions/amp-user-notification/0.1/amp-user-notification.js

```diff
@@ -0,0 +1,378 @@
+   * @private
+   * @return {!Promise}
+   */
+  postDismissEnpoint_() {
+    return xhrFor(this.win_).fetchJson(this.dismissHref_, {
+      method: 'POST',
+      credentials: 'include',
+      body: {
+        'elementId': this.elementId_,
+        'ampUserId': this.ampUserId_
+      }
+    });
+  }
+
+  /**
+   * Success handler for `postShowEndpoint_`.
+   * @param {!PostResponseMetadataDef}
+   * @return {!Promise<boolean>}
+   * @private
+   */
+  onPostShowEndpointSuccess_(data) {
+    assert(typeof data['showNotification'] == 'boolean', '`showNotification` ' +
+        'should be a boolean. Got "%s" which is of type %s.',
+        data['showNotification'], typeof data['showNotification']);
+
+    if (!data['showNotification']) {
+      // If no notification needs to be shown, resolve the `dialogPromise_`
+      // right away with false.
+      this.dialogResolve_();
+    }
+    return Promise.resolve(data['showNotification']);
+  }
+
+  /**
+   * Get async cid service.
+   * @return {!Promise}
+   * @private
+   */
+  getAsyncCid_() {
+    this.dialogPromise_ = new Promise(resolve => {
+      this.dialogResolve_ = resolve;
+    });
+    return cidFor(this.win_).then(cid => {
+      // `amp-user-notification` is our cid scope, while we give it a resolved
+      // promise for the 2nd argument so that the 3rd argument (the
+      // persistentConsent) is the one used to resolve getting
+      // the external CID.
+      // The dialogPromise_ is never rejected,
+      // the user only really has 1 option to accept/dismiss (to resolve)
+      // the notification or have the nagging notification sitting there
+      // (to never resolve).
+      return cid.get('amp-user-notification',
+          Promise.resolve(), this.dialogPromise_);
+    });
+  }
+
+  /** @override */
+  shouldShow() {
+    return this.getAsyncCid_()
+        .then(this.postShowEndpoint_.bind(this))
+        .then(this.onPostShowEndpointSuccess_.bind(this));
+  }
+
+  /** @override */
+  show() {
+    this.element.style.display = '';
+    this.element.classList.add('amp-active');
+    return this.dialogPromise_;
+  }
+
+  /**
+   * Hides the current user notification and invokes the `dialogResolve_`
+   * method. Removes the `.amp-active` class from the element.
+   */
+  dismiss() {
+    this.element.classList.remove('amp-active');
+    this.element.classList.add('amp-hidden');
+    this.dialogResolve_();
+    this.postDismissEnpoint_();
+  }
+}
+
+
+/**
+ * UserNotificationManager handles `amp-user-notification`
+ * queuing and registration, as well as exposing the components
+ * dismiss promise.
+ */
+export class UserNotificationManager {
+
+  constructor(window) {
+    this.win_ = window;
+
+    /** @private @const {!Object<!UserNotificationDeferDef>} */
+    this.deferRegistry_ = Object.create(null);
+
+    /** @private @const {!Viewer} */
+    this.viewer_ = viewerFor(this.win_);
+
+    /** @private @const {?function(*)} */
+    this.documentReadyResolve_ = null;
```

Comment:
This is a good example where we can actually avoid storing references to the resolver function. Consider this:

```
const documentReaderPromise = new Promise(resolve => {
  onDocumentReader(this.win_.document, resolve);
});

const viewerVisiblePromise = this.viewer_.whenVisible();

this.managerReaderPromise_ = promise.all(documentReaderPromise,
    viewerVisiblePromise);
```

That's it - no extra state stored.